[wasm][debugger] Fix debugger behavior when the type has ToString method overridden

thaystg opened the pull request:
The description of an object that has ToString implemented should bethe result of the ToString and not the type name.

Fixes #76748 

Using this PR:
![image](https://user-images.githubusercontent.com/4503299/194716463-281603dc-6ae4-4d05-ba63-28c0d639318c.png)

thaystg committed:
The description of an object that has ToString implemented should be the result of the ToString and not the type name.

thaystg committed:
Fix if cannot find ToString method.

thaystg committed:
removing my name of the code

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/JObjectValueCreator.cs:
```diff
@@ -293,6 +293,21 @@ private async Task<JObject> ReadAsObjectValue(MonoBinaryReader retDebuggerCmdRea
                 return Create(value: className, type: "symbol", description: className);
             }
         }
+        var typeInfo = await _sdbAgent.GetTypeInfo(type_id[0], token);
```
If a type has `ToString` overridden, and has a Debugger* attribute, then which one would be used?

Also, this won't work for ValueTypes, right?

radical commented on src/mono/wasm/debugger/tests/debugger-test/debugger-test.cs:
```diff
@@ -1371,4 +1371,17 @@ public static void CheckArguments(ReadOnlySpan<object> parameters)
+    {
+        return "hello";
+    }
+    public static void Run()
```
Can we also add a tests for:

```csharp
class A { public override string ToString() { ... } }
class B: A { .. }

var b = new B();
```
and
```csharp
class A { .. }
class B: A { public override string ToString() { .. } }

A a = new B();
```

radical replied:
Along with `A a = new B();`, `object o = new B();` would be useful too.

thaystg replied:
done

thaystg committed:
Implementing support for inherited classes with ToString overriden.
Implement support for value types with ToString overriden.
Fix Debugger* is more relevant than ToString

thaystg committed:
Fix test cases.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/ValueTypeClass.cs:
```diff
@@ -110,17 +111,39 @@ JObject GetFieldWithMetadata(FieldTypeClass field, JObject fieldValue, bool isSt
+            if (typeInfo.Name == "object")
+                return null;
+            Microsoft.WebAssembly.Diagnostics.MethodInfo methodInfo = typeInfo.Info.Methods.FirstOrDefault(m => m.Name == "ToString");
+            if (!IsEnum && methodInfo == null)
```
Should this be `||`?

radical replied:
Can this method exit early in case of `IsEnum==true`?

thaystg replied:
No should be `&&`, if it's a enum we want to call ToString even if we cannot find the method.

radical replied:
We don't get the method in case of an enum? Is that the case for other valuetypes? `record struct`?

thaystg replied:
Only for enum, I added tests for record also.

radical replied:
Are no methods returned for enums at all? Is that a debugger-agent thing?

thaystg replied:
No methods returned for enums at all, not sure if it's a debugger-agent thing or if they are really not implemented in the enum, they are somewhere like a "base class".

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/ValueTypeClass.cs:
```diff
@@ -110,17 +111,39 @@ JObject GetFieldWithMetadata(FieldTypeClass field, JObject fieldValue, bool isSt
+            if (methodIds == null)
+                return null;
+            try {
+                var retMethod = await sdbAgent.InvokeMethod(Buffer, methodIds[0], token, "methodRet");
```
There could be other `ToString` methods like too `public string ToString(int x) => $"{x+2}";`, so we need to find the one that has no parameters.

thaystg replied:
Done

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/ValueTypeClass.cs:
```diff
@@ -110,17 +111,39 @@ JObject GetFieldWithMetadata(FieldTypeClass field, JObject fieldValue, bool isSt
                 return fieldValue;
             }
         }
+        public async Task<string> ToString(MonoSDBHelper sdbAgent, CancellationToken token)
```
nit: rename to `InvokeToStringAsync`, or just `ToStringAsync`.

thaystg replied:
Done

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/ValueTypeClass.cs:
```diff
@@ -110,17 +111,39 @@ JObject GetFieldWithMetadata(FieldTypeClass field, JObject fieldValue, bool isSt
-                if (methodIds == null)
+                var toString = await ToString(sdbAgent, token);
+                if (toString == null)
                     throw new InternalErrorException($"Cannot find method 'ToString' on typeId = {TypeId}");
```
Should we log this, and fallback to some default value?

thaystg replied:
I just kept the old behavior that was throwing the exception, but I think we can, if you think it's better

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/ValueTypeClass.cs:
```diff
@@ -110,17 +111,39 @@ JObject GetFieldWithMetadata(FieldTypeClass field, JObject fieldValue, bool isSt
-                if (methodIds == null)
+                var toString = await ToString(sdbAgent, token);
+                if (toString == null)
                     throw new InternalErrorException($"Cannot find method 'ToString' on typeId = {TypeId}");
```
Also, the error isn't correct because `toString` could be null because we couldn't find it, or it wasn't applicable, or invoking the method failed with an exception.

thaystg replied:
changed.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/ValueTypeClass.cs:
```diff
@@ -110,17 +111,39 @@ JObject GetFieldWithMetadata(FieldTypeClass field, JObject fieldValue, bool isSt
                 return fieldValue;
             }
         }
+        public async Task<string> ToString(MonoSDBHelper sdbAgent, CancellationToken token)
```
```suggestion

        public async Task<string> ToString(MonoSDBHelper sdbAgent, CancellationToken token)
```

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/JObjectValueCreator.cs:
```diff
@@ -265,22 +265,51 @@ async Task<string> GetNullObjectClassName()
+            foreach (var typeId in typeIds)
+            {
+                var typeInfo = await _sdbAgent.GetTypeInfo(typeId, token);
+                if (typeInfo != null && typeInfo.Name != "object")
```
nit:
```suggestion
                if (typeInfo == null || typeInfo.Name == "object")
                    continue;
```

radical replied:
When could `typeInfo` be `null`? If it is null for a valid type, and that has a `ToString` override, then we will be skipping that, and using a parent's `ToString`.

And that would be a useful test to add - derived, and parent class both having `ToString` overridden, just to confirm that we are picking the correct one.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/JObjectValueCreator.cs:
```diff
@@ -265,22 +265,51 @@ async Task<string> GetNullObjectClassName()
+        string description = "";
+        try {
+            foreach (var typeId in typeIds)
+            {
```
Can we extract the body of the loop to `MonoSDBHelper`, and use that here, and in `ValueTypeClass`?

thaystg replied:
done

radical commented on src/mono/wasm/debugger/tests/debugger-test/debugger-test.cs:
```diff
@@ -1371,4 +1371,45 @@ public static void CheckArguments(ReadOnlySpan<object> parameters)
+        }
+    }
+
+    struct ToStringOverridenE 
```
```suggestion
    struct ToStringOverridenE
```

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/JObjectValueCreator.cs:
```diff
@@ -265,22 +265,51 @@ async Task<string> GetNullObjectClassName()
         var objectId = retDebuggerCmdReader.ReadInt32();
-        var type_id = await _sdbAgent.GetTypeIdsForObject(objectId, false, token);
-        string className = await _sdbAgent.GetTypeName(type_id[0], token);
+        var typeIds = await _sdbAgent.GetTypeIdsForObject(objectId, true, token);
```
use parameter name for the bool. And 👍 on the local rename!

thaystg replied:
done.

thaystg committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Adding more tests and addressing @radical comments

thaystg committed:
Adding tests for record

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/ValueTypeClass.cs:
```diff
@@ -111,16 +112,46 @@ JObject GetFieldWithMetadata(FieldTypeClass field, JObject fieldValue, bool isSt
         public async Task<JObject> ToJObject(MonoSDBHelper sdbAgent, bool forDebuggerDisplayAttribute, CancellationToken token)
         {
             string description = className;
             if (ShouldAutoInvokeToString(className) || IsEnum)
```
If a type has a DebuggerDisplayAttribute, and has ToString overridden, then what takes precedence?

thaystg replied:
If it's not a ShouldAutoInvode or a Enum, then DebuggerDisplayAttribute is more important.

thaystg committed:
Moving implementation of InvokeToStringAsync to sdbhelper and reuse it on objects and valuetypes.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/ValueTypeClass.cs:
```diff
@@ -116,11 +117,11 @@ public async Task<JObject> ToJObject(MonoSDBHelper sdbAgent, bool forDebuggerDis
-                    throw new InternalErrorException($"Cannot find method 'ToString' on typeId = {TypeId}");
-                var retMethod = await sdbAgent.InvokeMethod(Buffer, methodIds[0], token, "methodRet");
-                description = retMethod["value"]?["value"].Value<string>();
+                var toString = await sdbAgent.InvokeToStringAsync(new List<int>(){TypeId}, true, IsEnum, Id.Value, IsEnum ? BindingFlags.Default : BindingFlags.DeclaredOnly, token);
```
```suggestion
                var toString = await sdbAgent.InvokeToStringAsync(new int[]{ TypeId }, true, IsEnum, Id.Value, IsEnum ? BindingFlags.Default : BindingFlags.DeclaredOnly, token);
```

.. and named arguments, because of the two bools.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/ValueTypeClass.cs:
```diff
@@ -129,6 +130,12 @@ public async Task<JObject> ToJObject(MonoSDBHelper sdbAgent, bool forDebuggerDis
                     description = displayString;
+                else
+                {
+                    var toString = await sdbAgent.InvokeToStringAsync(new List<int>(){TypeId}, true, IsEnum, Id.Value, IsEnum ? BindingFlags.Default : BindingFlags.DeclaredOnly, token);
```
```suggestion
                    var toString = await sdbAgent.InvokeToStringAsync(new int[]{ TypeId }, true, IsEnum, Id.Value, IsEnum ? BindingFlags.Default : BindingFlags.DeclaredOnly, token);
```

.. and named arguments.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs:
```diff
@@ -1839,6 +1839,43 @@ public Task<JObject> InvokeMethod(DotnetObjectId dotnetObjectId, CancellationTok
+                foreach (var typeId in typeIds)
+                {
+                    var typeInfo = await GetTypeInfo(typeId, token);
+                    if (typeInfo == null)
+                        continue;
+                    if (typeInfo.Name == "object")
+                        continue;
```
nit:
```suggestion
                    if (typeInfo == null || typeInfo.Name == "object")
                        continue;
```

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/ValueTypeClass.cs:
```diff
@@ -129,6 +130,12 @@ public async Task<JObject> ToJObject(MonoSDBHelper sdbAgent, bool forDebuggerDis
                 string displayString = await sdbAgent.GetValueFromDebuggerDisplayAttribute(Id, TypeId, token);
                 if (displayString != null)
                     description = displayString;
```
```suggestion
                if (displayString != null)
                {
                    description = displayString;
                }
```

radical reviewed (approved):
Other than the minor comments, LGTM 👍

thaystg committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Addressing @radical comments.

ilonatommy reviewed (approved):
lgtm, good job👍
